File: Assets/Script/Test/TextInput.lua
-- [yue]: Script/Test/TextInput.yue
local Class = Dora.Class -- 1
local Label = Dora.Label -- 1
local Vec2 = Dora.Vec2 -- 1
local Line = Dora.Line -- 1
local Color = Dora.Color -- 1
local loop = Dora.loop -- 1
local sleep = Dora.sleep -- 1
local math = _G.math -- 1
local ClipNode = Dora.ClipNode -- 1
local Size = Dora.Size -- 1
local Keyboard = Dora.Keyboard -- 1
local App = Dora.App -- 1
local Node = Dora.Node -- 1
local property = Dora.property -- 1
local LineRect = require("UI.View.Shape.LineRect") -- 3
local SolidRect = require("UI.View.Shape.SolidRect") -- 4
local utf8 = require("utf-8") -- 5
local TextInput = Class((function(args) -- 7
	local fontName, fontSize, width, height, hint, text = args.fontName, args.fontSize, args.width, args.height, args.hint, args.text -- 8
	if hint == nil then -- 13
		hint = "" -- 13
	end -- 13
	if text == nil then -- 14
		text = hint -- 14
	end -- 14
	local label -- 17
	do -- 17
		local _with_0 = Label(fontName, fontSize) -- 17
		_with_0.batched = false -- 18
		_with_0.text = text -- 19
		_with_0.y = height / 2 - fontSize / 2 -- 20
		_with_0.anchor = Vec2.zero -- 21
		_with_0.alignment = "Left" -- 22
		label = _with_0 -- 17
	end -- 17
	local cursor = Line({ -- 24
		Vec2.zero, -- 24
		Vec2(0, fontSize + 2) -- 24
	}, Color(0xffffffff)) -- 24
	local blink -- 25
	blink = function() -- 25
		return loop(function() -- 25
			cursor.visible = true -- 26
			sleep(0.5) -- 27
			cursor.visible = false -- 28
			return sleep(0.5) -- 29
		end) -- 29
	end -- 25
	cursor.y = label.y -- 31
	cursor.visible = false -- 32
	local updateText -- 34
	updateText = function(text) -- 34
		label.text = text -- 35
		local offsetX = math.max(label.width + 3 - width, 0) -- 36
		label.x = -offsetX -- 37
		cursor.x = label.width - offsetX - 10 -- 38
		return cursor:schedule(blink()) -- 39
	end -- 34
	local node -- 41
	do -- 41
		local _with_0 = ClipNode(SolidRect({ -- 41
			width = width, -- 41
			height = height -- 41
		})) -- 41
		local textEditing = "" -- 42
		local textDisplay = "" -- 43
		_with_0.size = Size(width, height) -- 45
		_with_0.position = Vec2(width, height) / 2 -- 46
		_with_0.hint = hint -- 47
		_with_0:addChild(label) -- 48
		_with_0:addChild(cursor) -- 49
		local updateIMEPos -- 51
		updateIMEPos = function(next) -- 51
			return _with_0:convertToWindowSpace(Vec2(-label.x + label.width, 0), function(pos) -- 52
				Keyboard:updateIMEPosHint(pos) -- 53
				if next then -- 54
					return next() -- 54
				end -- 54
			end) -- 52
		end -- 51
		local startEditing -- 55
		startEditing = function() -- 55
			return updateIMEPos(function() -- 56
				_with_0:detachIME() -- 57
				return _with_0:attachIME() -- 58
			end) -- 56
		end -- 55
		_with_0.updateDisplayText = function(_self, text) -- 59
			textDisplay = text -- 60
			label.text = text -- 61
		end -- 59
		_with_0:onAttachIME(function() -- 63
			_with_0.keyboardEnabled = true -- 64
			return updateText(textDisplay) -- 65
		end) -- 63
		_with_0:onDetachIME(function() -- 67
			_with_0.keyboardEnabled = false -- 68
			cursor.visible = false -- 69
			cursor:unschedule() -- 70
			textEditing = "" -- 71
			label.x = 0 -- 72
			if textDisplay == "" then -- 73
				label.text = _with_0.hint -- 73
			end -- 73
		end) -- 67
		_with_0:onTapped(function(touch) -- 75
			if touch.first then -- 75
				return startEditing() -- 75
			end -- 75
		end) -- 75
		_with_0:onKeyPressed(function(key) -- 77
			if App.platform == "Android" and utf8.len(textEditing) == 1 then -- 78
				if key == "BackSpace" then -- 79
					textEditing = "" -- 79
				end -- 79
			else -- 81
				if textEditing ~= "" then -- 81
					return -- 81
				end -- 81
			end -- 78
			if "BackSpace" == key then -- 83
				if #textDisplay > 0 then -- 84
					textDisplay = utf8.sub(textDisplay, 1, -2) -- 85
					return updateText(textDisplay) -- 86
				end -- 84
			elseif "Return" == key then -- 87
				return _with_0:detachIME() -- 88
			end -- 88
		end) -- 77
		_with_0:onTextInput(function(text) -- 90
			textDisplay = utf8.sub(textDisplay, 1, -1 - utf8.len(textEditing)) .. text -- 91
			textEditing = "" -- 92
			updateText(textDisplay) -- 93
			return updateIMEPos() -- 94
		end) -- 90
		_with_0:onTextEditing(function(text, start) -- 96
			textDisplay = utf8.sub(textDisplay, 1, -1 - utf8.len(textEditing)) .. text -- 97
			textEditing = text -- 98
			label.text = textDisplay -- 99
			local offsetX = math.max(label.width + 3 - width, 0) -- 100
			label.x = -offsetX -- 101
			local charSprite = label:getCharacter(utf8.len(textDisplay) - utf8.len(textEditing) + start) -- 102
			if charSprite then -- 103
				cursor.x = charSprite.x + charSprite.width / 2 - offsetX + 1 -- 104
				cursor:schedule(blink()) -- 105
			else -- 107
				updateText(textDisplay) -- 107
			end -- 103
			return updateIMEPos() -- 108
		end) -- 96
		node = _with_0 -- 41
	end -- 41
	local _with_0 = Node() -- 110
	_with_0.content = node -- 111
	_with_0.cursor = cursor -- 112
	_with_0.label = label -- 113
	_with_0.size = Size(width, height) -- 114
	_with_0:addChild(node) -- 115
	return _with_0 -- 110
end), { -- 117
	text = property((function(self) -- 117
		return self.label.text -- 117
	end), function(self, value) -- 118
		self.content:detachIME() -- 119
		return self.content:updateDisplayText(value) -- 120
	end) -- 117
}) -- 7
local _with_0 = TextInput({ -- 124
	hint = "点这里进行输入", -- 124
	width = 300, -- 125
	height = 60, -- 126
	fontName = "sarasa-mono-sc-regular", -- 127
	fontSize = 40 -- 128
}) -- 123
local themeColor = App.themeColor:toARGB() -- 130
_with_0.label.color = Color(0xffff0080) -- 133
_with_0:addChild(LineRect({ -- 135
	x = -2, -- 135
	width = _with_0.width + 4, -- 136
	height = _with_0.height, -- 137
	color = themeColor -- 138
})) -- 134
return _with_0 -- 123

File: Assets/Script/Dev/Mobile/TextInput.ts
import { App, ClipNode, Color, Color3, DrawNode, Keyboard, KeyName, Label, Node, Size, sleep, TextAlign, thread, Vec2 } from "Dora";
import * as nvg from "nvg";

export const inputLength = (text: string) => utf8.len(text)[0] ?? 0;
export const inputSlice = (text: string, start: number, end = inputLength(text)) => {
	const first = utf8.offset(text, start + 1) ?? text.length + 1;
	const last = (utf8.offset(text, end + 1) ?? text.length + 1) - 1;
	return string.sub(text, first, last);
};
export const insertInputText = (text: string, at: number, inserted: string) => inputSlice(text, 0, at) + inserted + inputSlice(text, at);

// Explicit wrapping preserves whitespace and a source-index -> caret mapping.
// Sarasa is monospaced; widths come from the same engine font, not byte counts.
export function layoutInput(text: string, width: number, advance: (this: void, char: string) => number) {
	let x = 0, row = 0;
	const stops = [{ x: 0, row: 0 }];
	const display: string[] = [];
	for (const [, code] of utf8.codes(text)) {
		const char = utf8.char(code);
		if (char === "\n") { display.push(char); row++; x = 0; }
		else {
			const step = advance(char);
			if (x > 0 && x + step > width) {
				display.push("\n"); row++; x = 0;
				stops[stops.length - 1] = { x, row };
			}
			display.push(char); x += step;
		}
		stops.push({ x, row });
	}
	return { text: display.join(""), stops, rows: row + 1 };
}

export function createTextInputView(input: Node.Type, fontSize: number, singleLine = false, background = 0xff171c26) {
	const border = Node(); border.tag = `${input.tag}-border`;
	border.color3 = Color3(0x343b48); input.addChild(border, -1);
	input.onRender(() => {
		nvg.Save(); nvg.ApplyTransform(input);
		nvg.BeginPath(); nvg.RoundedRect(0, 0, input.width, input.height, 12);
		nvg.FillPaint(nvg.LinearGradient(0, input.height, 0, 0, Color(0xff242d3b), Color(background)));
		nvg.Fill();
		nvg.BeginPath(); nvg.RoundedRect(0.75, 0.75, input.width - 1.5, input.height - 1.5, 11.25);
		nvg.StrokeWidth(1.5); nvg.StrokeColor(Color(border.color3)); nvg.Stroke(); nvg.Restore();
		return false;
	});
	const insetX = 12, insetY = 8;
	const width = math.max(1, input.width - insetX * 2 - 2);
	const height = math.max(1, input.height - insetY * 2);
	const lineHeight = fontSize + 4;
	const makeLabel = () => {
		const label = Label("sarasa-mono-sc-regular", fontSize, true);
		if (!label) throw new Error("Missing mobile input font");
		label.alignment = TextAlign.Left; label.anchor = Vec2(0, 1); label.textWidth = -1;
		return label;
	};
	// Unparented Dora nodes are auto-attached to Director.entry on the next frame.
	// Own the metrics probe under the input so it never renders or outlives the UI.
	const measure = makeLabel(); measure.tag = "remix-input-measure"; measure.visible = false;
	input.addChild(measure);
	measure.batched = false; measure.text = "M";
	// Lua-facing getCharacter is 1-based (the C++ method itself is 0-based).
	const markerX = measure.getCharacter(1)?.x ?? 0;
	const singleHeight = measure.height;
	measure.text = "M\nM";
	const gap = measure.lineGap + lineHeight - (measure.height - singleHeight);
	const widths: Record<string, number> = {};
	const advance = (char: string) => {
		if (widths[char] !== undefined) return widths[char];
		measure.text = char + "M";
		const step = math.max(1, (measure.getCharacter(2)?.x ?? markerX + fontSize) - markerX);
		widths[char] = step;
		return step;
	};
	const stencil = DrawNode();
	stencil.drawPolygon([Vec2.zero, Vec2(width + 2, 0), Vec2(width + 2, height), Vec2(0, height)], Color(0xffffffff));
	const clip = ClipNode(stencil); clip.tag = "remix-input-clip";
	clip.anchor = Vec2.zero; clip.position = Vec2(insetX, insetY); clip.size = Size(width + 2, height);
	input.addChild(clip, 1);
	const content = Node(); content.tag = "remix-input-content"; clip.addChild(content);
	const label = makeLabel(); label.tag = "remix-input-text"; label.lineGap = gap; label.color3 = Color3(0xf4f1e8); content.addChild(label);
	const placeholder = makeLabel(); placeholder.tag = "remix-input-placeholder"; placeholder.lineGap = gap;
	const textTop = singleLine ? (height + singleHeight) / 2 : height;
	placeholder.textWidth = singleLine ? -1 : width; placeholder.y = textTop; placeholder.color3 = Color3(0xa8afbd); clip.addChild(placeholder);
	const caret = DrawNode(); caret.tag = "remix-input-caret";
	caret.drawPolygon([Vec2.zero, Vec2(1, 0), Vec2(1, fontSize + 2), Vec2(0, fontSize + 2)], Color(0xffffcc33));
	content.addChild(caret);
	let layout = layoutInput("", width, advance);
	let lastText = "", active = false, blink = 0, offset = 0, index = 0;
	const maxOffset = () => math.max(0, singleLine ? layout.stops[layout.stops.length - 1].x - width : layout.rows * lineHeight - height);
	const position = () => {
		content.x = singleLine ? -offset : 0;
		content.y = singleLine ? 0 : offset; label.y = textTop;
		const stop = layout.stops[index];
		caret.position = Vec2(singleLine ? stop.x : math.min(width, stop.x), singleLine ? (height - fontSize - 2) / 2 : height - stop.row * lineHeight - fontSize - 2);
	};
	const follow = () => {
		const stop = layout.stops[index];
		if (singleLine) {
			if (stop.x < offset) offset = stop.x;
			if (stop.x > offset + width) offset = stop.x - width;
		} else {
			const top = stop.row * lineHeight;
			if (top < offset) offset = top;
			if (top + lineHeight > offset + height) offset = top + lineHeight - height;
		}
		offset = math.max(0, math.min(maxOffset(), offset)); position();
	};
	caret.schedule(dt => { blink += dt; caret.visible = active && (blink % 1 < 0.5 || App.reducedMotion); return false; });
	const nearest = (x: number, row: number) => {
		let closest = 0, distance = math.huge;
		layout.stops.forEach((stop, i) => {
			const d = math.abs(stop.row - row) * (width + 1) + math.abs(stop.x - x);
			if (d < distance) { distance = d; closest = i; }
		});
		return closest;
	};
	return {
		label, caret, placeholder, clip, border,
		update: (text: string, hint: string, focused: boolean, cursor: number) => {
			if (singleLine) text = string.gsub(text, "[\r\n]", "")[0];
			if (lastText !== text) { layout = layoutInput(text, singleLine ? math.huge : width, advance); label.text = layout.text; lastText = text; }
			placeholder.text = hint; placeholder.visible = text === "" && !focused;
			border.color3 = Color3(focused ? 0xffcc33 : 0x343b48);
			active = focused; index = math.max(0, math.min(layout.stops.length - 1, cursor));
			blink = 0; caret.visible = active; follow();
		},
		caretPosition: () => Vec2(insetX + caret.x + content.x, insetY + caret.y + content.y),
		indexAt: (point: Vec2.Type) => nearest(point.x - insetX - content.x, singleLine ? 0 : math.max(0, math.floor((height + offset - (point.y - insetY)) / lineHeight))),
		verticalIndex: (cursor: number, direction: number) => { const stop = layout.stops[math.min(cursor, layout.stops.length - 1)]; return nearest(stop.x, stop.row + direction); },
		scroll: (delta: number) => { offset = math.max(0, math.min(maxOffset(), offset + delta)); position(); },
	};
}

export interface TextInputOptions {
	fontSize: number;
	singleLine?: boolean;
	background?: number;
	isSecure?(): boolean;
	getText(this: void): string;
	setText(this: void, text: string): void;
	getPlaceholder(this: void): string;
	isEnabled(this: void): boolean;
	// Return true to consume Return; otherwise multiline inputs insert a newline.
	onReturn?(this: void, modified: boolean): boolean;
}

// Shared editing/focus lifecycle. Callers own business state, not IME internals.
export function createTextInput(options: TextInputOptions) {
	let node: Node.Type | undefined;
	let view: ReturnType<typeof createTextInputView> | undefined;
	let focused = false, composition = "", compositionCursor = 0, cursor = 0, revision = 0, dragDistance = 0;
	const normalize = (text: string) => options.singleLine ? string.gsub(text, "[\r\n]", "")[0]
		: string.gsub(string.gsub(text, "\r\n", "\n")[0], "\r", "\n")[0];
	const updateIMEPos = (next?: (this: void) => void) => {
		const target = node;
		if (!target || !view) return;
		const captured = revision;
		const caret = view.caretPosition();
		target.convertToWindowSpace(Vec2(math.max(12, math.min(target.width - 12, caret.x)), math.max(8, math.min(target.height - 8, caret.y))), pos => {
			if (node !== target || captured !== revision || !options.isEnabled()) return;
			Keyboard.updateIMEPosHint(pos); next?.();
		});
	};
	const refresh = () => {
		const text = options.getText();
		cursor = math.min(cursor, inputLength(text));
		const editing = insertInputText(text, cursor, composition);
		const display = options.isSecure?.() ? string.rep("•", inputLength(editing)) : editing;
		view?.update(display, options.getPlaceholder(), focused, cursor + compositionCursor);
		if (focused) updateIMEPos();
	};
	const clearFocus = () => {
		revision++; focused = false; composition = ""; compositionCursor = 0;
		if (node) node.keyboardEnabled = false;
		refresh();
	};
	const blur = () => { if (focused) node?.detachIME(); clearFocus(); };
	const focus = (reopen = true) => {
		if (!options.isEnabled()) return;
		revision++;
		updateIMEPos(() => {
			if (reopen) node?.detachIME();
			node?.attachIME(); updateIMEPos();
		});
	};
	const setValue = (text: string, at: number) => { options.setText(text); cursor = at; refresh(); };
	const textInput = (text: string) => {
		if (!options.isEnabled()) return;
		composition = ""; compositionCursor = 0;
		const value = normalize(text);
		setValue(insertInputText(options.getText(), cursor, value), cursor + inputLength(value));
	};
	const keyInput = (key: KeyName) => {
		if (!options.isEnabled()) return;
		if (key === KeyName.Escape) { blur(); return; }
		if (composition !== "") return;
		const value = options.getText();
		if (key === KeyName.BackSpace && cursor > 0) setValue(inputSlice(value, 0, cursor - 1) + inputSlice(value, cursor), cursor - 1);
		else if (key === KeyName.Delete && cursor < inputLength(value)) setValue(inputSlice(value, 0, cursor) + inputSlice(value, cursor + 1), cursor);
		else if (key === KeyName.Home || key === KeyName.End || key === KeyName.Left || key === KeyName.Right || key === KeyName.Up || key === KeyName.Down) {
			cursor = key === KeyName.Home ? 0 : key === KeyName.End ? inputLength(value)
				: key === KeyName.Up || key === KeyName.Down ? view?.verticalIndex(cursor, key === KeyName.Up ? -1 : 1) ?? cursor
				: math.max(0, math.min(inputLength(value), cursor + (key === KeyName.Left ? -1 : 1)));
			refresh();
		} else if (key === KeyName.Return) {
			const modified = Keyboard.isKeyPressed(KeyName.LCtrl) || Keyboard.isKeyPressed(KeyName.RCtrl) || Keyboard.isKeyPressed(KeyName.LGui) || Keyboard.isKeyPressed(KeyName.RGui);
			if (!options.onReturn?.(modified) && !options.singleLine) textInput("\n");
		}
	};
	const unmount = () => { blur(); node = undefined; view = undefined; };
	return {
		refresh, focus, blur, unmount,
		pasteFromClipboard: (replace = false) => {
			if (!options.isEnabled()) return false;
			const value = normalize(App.getClipboardText());
			if (value === "") return false;
			if (replace) setValue(value, inputLength(value));
			else textInput(value);
			return true;
		},
		isFocused: () => focused,
		isComposing: () => composition !== "",
		deferFocus: () => {
			const captured = revision;
			thread(() => { sleep(0); if (captured === revision) focus(); });
		},
		mount: (target: Node.Type) => {
			unmount(); node = target;
			view = createTextInputView(target, options.fontSize, options.singleLine, options.background);
			target.touchEnabled = true; target.swallowTouches = true;
			target.onAttachIME(() => { focused = true; composition = ""; compositionCursor = 0; target.keyboardEnabled = true; refresh(); });
			target.onDetachIME(clearFocus);
			target.onTextInput(textInput);
			target.onTextEditing((text, start) => {
				if (!options.isEnabled()) return;
				composition = normalize(text); compositionCursor = math.max(0, math.min(inputLength(composition), start ?? inputLength(composition))); refresh();
			});
			target.onKeyDown(keyInput); target.keyboardEnabled = false;
			target.onTapBegan(() => { dragDistance = 0; });
			target.onTapMoved(touch => {
				if (!options.isEnabled()) return;
				dragDistance += math.abs(touch.delta.x) + math.abs(touch.delta.y);
				view?.scroll(options.singleLine ? -touch.delta.x : touch.delta.y); if (focused) updateIMEPos();
			});
			target.onMouseWheel(delta => { if (!options.isEnabled()) return; view?.scroll(-delta.y * 20); if (focused) updateIMEPos(); });
			target.onTapped(touch => {
				if (dragDistance > 5 || !options.isEnabled()) return;
				if (touch !== undefined && composition === "") cursor = view?.indexAt(touch.location) ?? cursor;
				refresh(); if (!focused) focus();
			});
			target.onCleanup(() => { if (node === target) unmount(); });
			refresh();
		},
	};
}
